NXP drivers: spi: spi_mcux_lpspi: enable CONFIG_SPI_ASYNC mode with DMA

DerekSnell opened the pull request:
Updates LPSPI driver to support async mode with DMA.  Enables `spi_loopback` testing async mode on all the NXP boards with the LPSPI peripherals.  Resolves #74907.

spi_loopback test with `CONFIG_SPI_ASYNC` passes on the following boards.  The test failed on 1 board which needs more investigation, a separate issue created for that board at https://github.com/zephyrproject-rtos/zephyr/issues/78238.
- Mimxrt1010_evk
- Mimxrt1020_evk
- Mimxrt1024_evk
- Mimxrt1040_evk
- Mimxrt1170_evk_cm7
- Frdm_ke17z512
- frdm_mcxn947//cpu0
- frdm_mcxn236

Commit:
drivers: spi: spi_mcux_lpspi: disable unused scatter/gather mode

DMA scatter/gather mode was unused in this driver.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
tests: drivers: spi_loopback: select CONFIG_SPI_ASYNC on NXP LPSPI boards

Enable testing async mode on all the NXP boards with the LPSPI peripherals.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

decsny commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -449,9 +450,9 @@ static int spi_rx_every_4(struct spi_dt_spec *spec)
+	const uint32_t tx_buf_size = TX_BUF_SIZE;
 
-	BUILD_ASSERT(tx_buf_size < BUF_SIZE,
+	BUILD_ASSERT(TX_BUF_SIZE < BUF_SIZE,
```
maybe move this build assert to be right after the macro definitions instead of in this function

DerekSnell replied:
good idea, done

decsny commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -532,6 +532,12 @@ static int dma_mcux_edma_reload(const struct device *dev, uint32_t channel,
 
+	if (data->edma_handle.tcdPool == NULL) {
+		/* Not using scatter/gather mode, re-enable the hardware request */
+		EDMA_EnableChannelRequest(DEV_EDMA_HANDLE(dev, channel)->base,
```
can you give an explanation of "why" in the commit message at least, maybe also in the comment

DerekSnell replied:
added details in comment and commit message

decsny commented on drivers/spi/spi_mcux_lpspi.c:
```diff
@@ -325,23 +325,60 @@ static void spi_mcux_dma_callback(const struct device *dev, void *arg,
 
 		if (data->ctx.tx_len == 0 && data->ctx.rx_len == 0) {
 			spi_context_complete(&data->ctx, spi_dev, 0);
+		} else {
```
nit, just return from the if block, instead of indenting an else block

DerekSnell replied:
good point, done

decsny commented on drivers/spi/spi_mcux_lpspi.c:
```diff
@@ -325,23 +325,60 @@ static void spi_mcux_dma_callback(const struct device *dev, void *arg,
+			dma_size = MIN(data->ctx.tx_len, data->ctx.rx_len);
+
+			if (dma_size == 0) {
+				dma_size = MAX(data->ctx.tx_len, data->ctx.rx_len);
```
why is it not always max?

DerekSnell replied:
Hi @decsny ,
Making this change (always MAX) here in `spi_mcux_dma_callback()`, the `spi_loopback` test passes.  However, if I make this same change in `spi_mcux_dma_rxtx_load()`, the "half start" test fails.  I believe this existing code is keeping the RX and TX transfers equal size until one completes, and then it finishes the other.

decsny commented on drivers/spi/spi_mcux_lpspi.c:
```diff
@@ -325,23 +325,60 @@ static void spi_mcux_dma_callback(const struct device *dev, void *arg,
+			data->status_flags = 0U;
+
+			/* Restart the RX DMA */
+			int ret;
```
nit declare variables at top of scope block

DerekSnell replied:
done

decsny commented on drivers/spi/spi_mcux_lpspi.c:
```diff
@@ -325,23 +325,60 @@ static void spi_mcux_dma_callback(const struct device *dev, void *arg,
+
+			ret = dma_reload(dev, data->dma_rx.channel,
+				(uint32_t) data->dma_rx.dma_blk_cfg.source_address,
+				(uint32_t)data->ctx.rx_buf,
```
nit inconsistent spacing

DerekSnell replied:
fixed spacing

decsny commented on drivers/spi/spi_mcux_lpspi.c:
```diff
@@ -325,23 +325,60 @@ static void spi_mcux_dma_callback(const struct device *dev, void *arg,
+				dma_size);
+			if (ret != 0) {
+				LOG_ERR("RX dma_reload() failed with error %x.",
+								ret);
```
nit newline not needed (especially if you take the suggestion above to remove the level of indent)

DerekSnell replied:
removed new line

decsny commented on drivers/spi/spi_mcux_lpspi.c:
```diff
@@ -325,23 +325,60 @@ static void spi_mcux_dma_callback(const struct device *dev, void *arg,
+			/* Restart the TX DMA */
+			dma_reload(dev, data->dma_tx.channel,
+				(uint32_t)data->ctx.tx_buf,
+				(uint32_t) data->dma_tx.dma_blk_cfg.dest_address,
```
more inconsistent spacing

DerekSnell replied:
fixed spacing

decsny reviewed (commented):
in Kconfig.mcux_lpspi can you please add something like `depends on DMA` if `SPI_ASYNC` to the LPSPI driver kconfig

Commit:
tests: drivers: spi_loopback: fix tx_buf_size compiler error

With some compiler optimizations, compiler error occurs when trying to test
tx_buf_size in BUILD_ASSERT.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
drivers: spi: spi_mcux_lpspi: select CONFIG_SPI_MCUX_LPSPI_DMA

If CONFIG_SPI_ASYNC, select the DMA feature

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
drivers: dma: dma_mcux_edma: enable HW request in dma_mcux_edma_reload()



If not using scatter/gather mode, re-enable the hardware channel requests.
The fsl_edma driver uses the eDMA DREQ bit to automatically disable the
hardware DMA request at the end of the major loop.



Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
drivers: spi: spi_mcux_lpspi: enable CONFIG_SPI_ASYNC mode with DMA

add DMA async support.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

danieldegrasse commented on drivers/spi/spi_mcux_lpspi.c:
```diff
@@ -322,26 +322,64 @@ static void spi_mcux_dma_callback(const struct device *dev, void *arg,
+		}
+
+		/* Restart the TX DMA */
+		ret = dma_reload(dev,
```
Have we verified this works correctly when the SPI is only sending or only receiving data? If memory serves, we usually configure the TX DMA or RX DMA with a step size of zero and a dummy buffer in these cases, so that DMA will run on both channels. My concern is that if we (for example) are sending TX data and not receiving anything, the LPSPI RX FIFO will overflow since the DMA will not be reading from it.

LukasTaroza commented:
Hello, I'm testing with mimxrt1060_evk. The tests are passing, but the functionality is incorrect. I am getting chip select for every byte when I should only get chip select for one block.
#78732 PR adds continuous mode (base->TCR |= LPSPI_TCR_CONT_MASK), which completely breaks dma spi (no test pass, chip select is never deselected). With this mode additional command end dma block should be sent to deselect chip select. I could add more information if needed.

decsny commented:
> #78732 PR adds continuous mode (base->TCR |= LPSPI_TCR_CONT_MASK), which completely breaks dma spi (no test pass, chip select is never deselected). With this mode additional command end dma block should be sent to deselect chip select. I could add more information if needed.

what do you mean by this, are you saying that PR breaks it? because if so, that's not my observations.

LukasTaroza commented:
I'm using main branch of zephyr, and executing `west build -p always -b mimxrt1060_evk -t flash tests/drivers/spi/spi_loopback` the results are: 
```
*** Booting Zephyr OS build v3.7.0-4815-g6773f3344575 ***                                                                                                                                       
Running TESTSUITE spi_loopback                                                                                                                                                                  
===================================================================                                                                                                                             
START - test_spi_loopback                                                                                                                                                                       
I: SPI test on buffers TX/RX 0x800040a0/0x80004080, frame size = 8, DMA enabled                                                                                                                 
I: SPI test slow config                                                                                                                                                                         
I: Start complete multiple                                                                                                                                                                      
E: Timeout waiting for transfer complete                                                                                                                                                        
E: Code -116                                                                                                                                                                                    
                                                                                                                                                                                                
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:142: spi_complete_multiple: (ret is true)                                                                   
SPI transceive failed                                                                                                                                                                           
 FAIL - test_spi_loopback in 0.231 seconds                                                                                                                                                      
===================================================================                                                                                                                             
TESTSUITE spi_loopback failed.                                                                                                                                                                  
                                                                                                                                                                                                
------ TESTSUITE SUMMARY START ------                                                                                                                                                           
                                                                                                                                                                                                
SUITE FAIL -   0.00% [spi_loopback]: pass = 0, fail = 1, skip = 0, total = 1 duration = 0.231 seconds                                                                                           
 - FAIL - [spi_loopback.test_spi_loopback] duration = 0.231 seconds                                                                                                                             
                                                                                                                                                                                                
------ TESTSUITE SUMMARY END ------                                                                                                                                                             
                                                                                                                                                                                                
===================================================================
PROJECT EXECUTION FAILED
```
![image](https://github.com/user-attachments/assets/8bc33ad1-971a-40dc-846e-a98337a8c816)

If I remove the `base->TCR |= LPSPI_TCR_CONT_MASK` in `trasceive_dma` function I get:
```
*** Booting Zephyr OS build v3.7.0-4815-g6773f3344575 ***                                                                                                                                       
Running TESTSUITE spi_loopback                                                                                                                                                                  
===================================================================                                                                                                                             
START - test_spi_loopback                                                                                                                                                                       
I: SPI test on buffers TX/RX 0x800040a0/0x80004080, frame size = 8, DMA enabled                                                                                                                 
I: SPI test slow config                                                                                                                                                                         
I: Start complete multiple                                                                                                                                                                      
I: Passed                                                                                                                                                                                       
I: Start complete loop                                                                                                                                                                          
I: Passed                                                                                                                                                                                       
I: Start null tx                                                                                                                                                                                
I: Passed                                                                                                                                                                                       
I: Start half start                                                                                                                                                                             
I: Passed
I: Start half end
I: Passed
I: Start every 4
I: Passed
I: Start rx bigger than tx
I: Passed
I: Start complete large transfers
I: Passed
I: SPI test fast config
I: Start complete multiple
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Start half end
I: Passed
I: Start every 4
I: Passed
I: Start rx bigger than tx
I: Passed
I: Start complete large transfers
I: Passed
I: SPI test fast config
I: Start complete multiple
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Start half end
I: Passed
I: Start every 4
I: Passed
I: Start rx bigger than tx
I: Passed
I: Start complete large transfers
I: Passed
I: Start complete loop
I: Passed
I: Start complete loop
I: Passed
I: All tx/rx passed
 PASS - test_spi_loopback in 0.205 seconds
===================================================================
TESTSUITE spi_loopback succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [spi_loopback]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.205 seconds
 - PASS - [spi_loopback.test_spi_loopback] duration = 0.205 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

![image](https://github.com/user-attachments/assets/a7ec125c-28cc-482b-a403-009045c9c0d1)

I think the flag is correct because otherwise, you could only send one byte commands. The documentation says that to use this function, you need to send extra command to deselect chip-select.
| CONT | Continuous transfer | Configures LPSPI for a continuous transfer that keeps PCS asserted between frames (as specified by FRAMESZ). You must write a new command word to cause PCS to negate. Also supports changing the command word at the frame size boundaries. |
|------|---------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

DerekSnell commented:
New PR https://github.com/zephyrproject-rtos/zephyr/pull/85668 replaces this one.
